src/python/grpcio/grpc/experimental/aio/_interceptor.py

```diff
@@ -325,14 +308,177 @@ async def debug_error_string(self) -> Optional[str]:
+    async def _invoke(self,
+                      interceptors: Sequence[UnaryUnaryClientInterceptor],
+                      method: bytes, timeout: Optional[float],
+                      metadata: Optional[MetadataType],
+                      credentials: Optional[grpc.CallCredentials],
+                      wait_for_ready: Optional[bool], request: RequestType,
+                      request_serializer: SerializingFunction,
+                      response_deserializer: DeserializingFunction
+                     ) -> UnaryStreamCall:
+        """Run the RPC call wrapped in interceptors"""
+
+        last_returned_call_from_interceptors = [None]
+
+        async def _run_interceptor(
+                interceptors: Iterator[UnaryStreamClientInterceptor],
+                client_call_details: ClientCallDetails,
+                request: RequestType,
+               ) -> _base_call.UnaryUnaryCall:
+
+            interceptor = next(interceptors, None)
+
+            if interceptor:
+                continuation = functools.partial(_run_interceptor, interceptors)
+
+                call_or_response_iterator = await interceptor.intercept_unary_stream(
+                    continuation, client_call_details, request)
+
+                if call_or_response_iterator is last_returned_call_from_interceptors[0]:
+                    return call_or_response_iterator
+                else:
+                    last_returned_call_from_interceptors[0] = UnaryStreamCallResponseIterator(
+                        last_returned_call_from_interceptors[0],
+                        call_or_response_iterator
+                    )
+                    return last_returned_call_from_interceptors[0]
+            else:
+                last_returned_call_from_interceptors[0] = UnaryStreamCall(
+                    request, _timeout_to_deadline(client_call_details.timeout),
+                    client_call_details.metadata,
+                    client_call_details.credentials,
+                    client_call_details.wait_for_ready, self._channel,
+                    client_call_details.method, request_serializer,
+                    response_deserializer, self._loop)
+
+                return last_returned_call_from_interceptors[0]
+
+        client_call_details = ClientCallDetails(method, timeout, metadata,
+                                                credentials, wait_for_ready)
+        return await _run_interceptor(iter(interceptors), client_call_details,
+                                      request)
+
+    async def _wait_for_interceptor_task_response_iterator(self) -> ResponseType:
         call = await self._interceptors_task
-        return await call.wait_for_connection()
+        async for response in call:
+            yield response
+
+    def __aiter__(self) -> AsyncIterable[ResponseType]:
+        return self._response_aiter
+
+    async def read(self) -> ResponseType:
+        return await self._response_aiter.asend(None)
```

Comment:
Question for you @lidizheng prohibiting mixing styles in the intercepted call is not gonna be straight forward. The interceptors only provide an iterator, hence the `read` needs to make the read from the iterator.

Technically we could still prohibit mix both usages by storing a kind of state in the intercepted call, but I would prefer to not have to replicate logic.

So my specific question would be, could I challenge the idea of having this prohibition in the main API? So allowing the user to use both primitives at the same time.

Indeed we could follow the same strategy as here, so having behind the scenes only one consumer that is the iterator which can be also consumed implicitly by the `read`.

Reply:
Sorry for the late reply. Last time our discussion around this topic ends with a conclusion that if we allow using both APIs, users might abuse them and writing racy code. Still, I think as a library we should prevent this behavior if possible.

On the other hand, the prohibition should be per-call-instance, so users should allow to use `read` for the real `UnaryStreamCall` and use either `__aiter__` or `read` on `InterceptedUnaryStreamCall`. Since there is a layer of abstraction, both `__aiter__` and `read` of  `InterceptedUnaryStreamCall` are actually calling `read` on `UnaryStreamCall`, it doesn't violate the API style mixing prohibition.

WDYT?

Reply:
Let me move the discussion to the final PR